pkg/aws/cloudmap.go

```diff
@@ -0,0 +1,351 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/klog"
+
+	awssdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/appmesh"
+	"github.com/aws/aws-sdk-go/service/servicediscovery"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	CreateServiceTimeout       = 10
+	DeregisterInstanceTimeout  = 10
+	GetServiceTimeout          = 10
+	ListInstancesPagesTimeout  = 10
+	ListNamespacesPagesTimeout = 10
+	ListServicesPagesTimeout   = 10
+	RegisterInstanceTimeout    = 10
+
+	//AttrAwsInstanceIPV4 is a special attribute expected by CloudMap.
+	//See https://github.com/aws/aws-sdk-go/blob/fd304fe4cb2ea1027e7fc7e21062beb768915fcc/service/servicediscovery/api.go#L5161
+	AttrAwsInstanceIPV4 = "AWS_INSTANCE_IPV4"
+	//AttrK8sPod is a custom attribute injected by app-mesh controller
+	AttrK8sPod = "k8s.io/pod"
+	//AttrK8sNamespace is a custom attribute injected by app-mesh controller
+	AttrK8sNamespace = "k8s.io/namespace"
+)
+
+//CloudMapAPI is wrapper util to invoke CloudMap API
+type CloudMapAPI interface {
+	CloudMapCreateService(context.Context, *appmesh.AwsCloudMapServiceDiscovery, string) (*CloudMapServiceSummary, error)
+	CloudMapGetService(context.Context, string) (*CloudMapServiceSummary, error)
+	RegisterInstance(context.Context, string, *corev1.Pod, *appmesh.AwsCloudMapServiceDiscovery) error
+	DeregisterInstance(context.Context, string, *appmesh.AwsCloudMapServiceDiscovery) error
+	ListInstances(context.Context, *appmesh.AwsCloudMapServiceDiscovery) ([]*servicediscovery.InstanceSummary, error)
+}
+
+//CloudMapCreateService calls AWS ServiceDiscovery CreateService API
+func (c *Cloud) CloudMapCreateService(ctx context.Context, cloudmapConfig *appmesh.AwsCloudMapServiceDiscovery, creatorRequestID string) (*CloudMapServiceSummary, error) {
+	key := awssdk.StringValue(cloudmapConfig.ServiceName) + "@" + awssdk.StringValue(cloudmapConfig.NamespaceName)
+
+	existingItem, exists, _ := c.serviceIDCache.Get(&cloudmapServiceCacheItem{
+		key: key,
+	})
+	if exists && existingItem != nil {
+		return &(existingItem.(*cloudmapServiceCacheItem)).value, nil
+	}
+
+	namespaceSummary, err := c.getNamespace(ctx, cloudmapConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if namespaceSummary == nil {
+		return nil, fmt.Errorf("Could not find namespace in cloudmap with name %s", awssdk.StringValue(cloudmapConfig.NamespaceName))
+	}
+
+	createServiceInput := &servicediscovery.CreateServiceInput{
+		CreatorRequestId: awssdk.String(creatorRequestID),
+		Name:             cloudmapConfig.ServiceName,
+		DnsConfig: &servicediscovery.DnsConfig{
+			NamespaceId:   awssdk.String(namespaceSummary.NamespaceID),
+			RoutingPolicy: awssdk.String(servicediscovery.RoutingPolicyMultivalue),
+			DnsRecords: []*servicediscovery.DnsRecord{
```

Comment:
I think we need to add this conditionally, I am getting this error: 
```
E0903 15:44:25.848882       1 controller.go:443] error syncing 'appmesh-demo/colorteller-red': Error handling cloudmap service discovery for virtual node colorteller-red-appmesh-demo: InvalidInput: When you create a service using a namespace that has a type of HTTP, you can't include a DnsConfig element.
        status code: 400, request id: 8c76f6af-1c79-4aa7-a24a-a2b2c3de0e5b
```

Reply:
HTTP namespace will not work with AppMesh because those service-names cannot be resolved by VPC DNS. In the current setup, application container still resolves the name using DNS resolver which gives an endpoint that is ignored by Envoy and further resolved. https://github.com/envoyproxy/envoy/issues/6748 will help in avoiding DNS altogether in future. Until then we should be using PrivateDnsNamespace.

Reply:
Ok we can call this out as its a little confusing.  Arguably, though, as long as the application gets some DNS response, then does everything work fine?  Whether its kube dns, route53, or some custom DNS solution the customer is running.

Reply:
@nckturner
> as long as the application gets some DNS response, then does everything work fine
Yeah that is case right now. We're also looking into adding some work-around we can give customers since some people may not be allowed to create a PrivateDnsNamespace.

Reply:
Updated PR with better error message when provided namespace is non DNS_PRIVATE. Marked it with a TODO to support HTTP namespaces. That should unblock this PR.